File: src/number/oocNumber.r
#ifndef OOC_NUMBER_R_
#define OOC_NUMBER_R_

#include "oocObject.r"

typedef struct OOC_Number OOC_Number;
typedef struct OOC_NumberClass OOC_NumberClass;

struct OOC_Number {
    OOC_Object object;
};

struct OOC_NumberClass {
    OOC_Class class;
    int            (*intValue)(const void* self);
    unsigned int   (*uintValue)(const void* self);
    long           (*longValue)(const void* self);
    unsigned long  (*ulongValue)(const void* self);
    float          (*floatValue)(const void* self);
    double         (*doubleValue)(const void* self);
};

#define OOC_NUMBER_METHOD_INT_VALUE    offsetof(OOC_NumberClass, intValue)
#define OOC_NUMBER_METHOD_UINT_VALUE   offsetof(OOC_NumberClass, uintValue)
#define OOC_NUMBER_METHOD_LONG_VALUE   offsetof(OOC_NumberClass, longValue)
#define OOC_NUMBER_METHOD_ULONG_VALUE  offsetof(OOC_NumberClass, ulongValue)
#define OOC_NUMBER_METHOD_FLOAT_VALUE  offsetof(OOC_NumberClass, floatValue)
#define OOC_NUMBER_METHOD_DOUBLE_VALUE offsetof(OOC_NumberClass, doubleValue)

#endif

File: src/number/oocNumber.c
#include "oocNumber.h"
#include "oocNumber.r"
#include "oocObject.h"

static OOC_NumberClass* NumberClass;
static OOC_NumberClass NumberClassInstance;

int ooc_numberIntValue(const void* self) {
    OOC_TYPE_CHECK(self, ooc_numberClass(), 0);
    OOC_NumberClass* c = (OOC_NumberClass*)ooc_classOf(self);
    return c->intValue(self);
}

unsigned int ooc_numberUIntValue(const void* self) {
    OOC_TYPE_CHECK(self, ooc_numberClass(), 0);
    OOC_NumberClass* c = (OOC_NumberClass*)ooc_classOf(self);
    return c->uintValue(self);
}

long ooc_numberLongValue(const void* self) {
    OOC_TYPE_CHECK(self, ooc_numberClass(), 0);
    OOC_NumberClass* c = (OOC_NumberClass*)ooc_classOf(self);
    return c->longValue(self);
}

unsigned long ooc_numberULongValue(const void* self) {
    OOC_TYPE_CHECK(self, ooc_numberClass(), 0);
    OOC_NumberClass* c = (OOC_NumberClass*)ooc_classOf(self);
    return c->ulongValue(self);
}

float ooc_numberFloatValue(const void* self) {
    OOC_TYPE_CHECK(self, ooc_numberClass(), 0);
    OOC_NumberClass* c = (OOC_NumberClass*)ooc_classOf(self);
    return c->floatValue(self);
}

double ooc_numberDoubleValue(const void* self) {
    OOC_TYPE_CHECK(self, ooc_numberClass(), 0);
    OOC_NumberClass* c = (OOC_NumberClass*)ooc_classOf(self);
    return c->doubleValue(self);
}

static void* ooc_numberClassInit(void) {
    if (ooc_classNew(&NumberClassInstance,
                    "Number",
                    sizeof(OOC_Number),
                    sizeof(OOC_NumberClass),
                    ooc_objectClass(),
                    OOC_MODIFIER_ABSTRACT,
                    0) != OOC_ERROR_NONE) {
        ooc_classDestroy(&NumberClassInstance);
        return NULL;
    }
    return &NumberClassInstance;
}

void* ooc_numberClass(void) {
    if (!NumberClass) {
        NumberClass = ooc_numberClassInit();
    }
    return NumberClass;
}
